### PR TITLE
Shape fix

### LINKED ISSUE #1524: InferShapeOfOutput returns a wrong shape.
This is triggered by #1503, but I managed to reproduce this on head. We haven't tested customizing allocation domains of segment boundaries, a fairly uncommon case until #1503. 

To reproduce this, 

```
git fetch origin infer_shape_bug
git checkout infer_shape_bug
python setup.py develop --debug && bin/nvfuser_tests --gtest_filter=AllocationDomainTest.Bug
```

I guess the root cause is that https://github.com/NVIDIA/Fuser/blob/582f29583ce337191c4d273781994ad03b121aee/csrc/executor.cpp#L589 should populate `symbolic_sizes` in rfactor order not allocation?

### PR DESCRIPTION
**Pull request by jjsjann123**
Fixes #1524 

Moves `transformOutputFromAllocationToRFactor` from `allocateOutput` to `inferShapeOfOutput` so intermediate tensor between segments would be allocated in the correct memory format.

### CODE REVIEW

**Comment by jjsjann123** [comment]
!build

**Comment by jjsjann123** [comment]
!build

**Comment by jjsjann123** [comment]
!build

File: csrc/executor.cpp [hunk 1]
```diff
@@ -908,6 +877,51 @@ at::Tensor transformOutputFromAllocationToRFactor(
...
+          id->toString());
+      expand_flags.push_back(true);
+    } else {
+      expand_flags.push_back(false);
+    }
+  }
+
+  auto pair_size_stride =
+      inferShape(tv, symbolic_sizes, expand_flags, expr_eval);
+  if (tv->hasAllocation()) {
```
Comment (wujingyue) [inline csrc/executor.cpp:910]: I'd early exit to reduce the def-use chain of `pair_size_stride`. 

```
if (!tv->hasAllocation()) {
  return pair_size_stride;
}
...
```

File: csrc/executor.cpp [hunk 2]
```diff
@@ -908,6 +877,51 @@ at::Tensor transformOutputFromAllocationToRFactor(
...
+          id->isBroadcast(),
+          "Non-broadcast domain should not have an expanded extent: ",
+          id->toString());
+      expand_flags.push_back(true);
+    } else {
+      expand_flags.push_back(false);
+    }
+  }
+
+  auto pair_size_stride =
```
Comment (wujingyue) [inline csrc/executor.cpp:908]:
```suggestion
  auto sizes_strides =
```

**Review by wujingyue — APPROVED** [review]
LGTM with comments. Thanks for the quick fix! 

Do you want to copy my reproducer to this PR?

**Comment by jjsjann123** [comment]
!build

File: test/test_allocation_domain.cpp [hunk 3]
```diff
@@ -1312,4 +1312,25 @@ TEST_F(AllocationDomainTest, TrivialStrideOrderTensorViewBuilder) {
   EXPECT_TRUE(!tv0->hasAllocation());
 }
 
+TEST_F(AllocationDomainTest, Bug) {
```
Comment (wujingyue) [inline test/test_allocation_domain.cpp:1315]:
```suggestion
TEST_F(AllocationDomainTest, Issue1524) {
```

**Review by wujingyue — APPROVED** [review]

**Comment by jjsjann123** [comment]
!build

**Comment by jjsjann123** [comment]
!build

**Pull request merged**